File: 151044081_hw8/BigramMap.java
import java.io.File;
import java.util.Scanner;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.BufferedReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.*;

public class BigramMap<T> implements Bigram<T> {
    
    private final int dataType ;
    
    private HashMap<Pair<T,T>,Integer> mymap;
    
    private Scanner input;
    
    private int countForNumGrams = 0;

    private Pair<T,T> [] peyir;

    private int capacity = 50 ;
    
    private ArrayList<T> array ;
    
    public BigramMap(int _dataType){
        
        this.dataType = _dataType;

        this.array = new ArrayList<>();

        this.mymap = new HashMap<>();

      //  this.peyir = new Pair[capacity];
    }
    
    public static boolean isInteger(String s) {
        
        try { 
            Integer.parseInt(s); 
        } 
		catch(NumberFormatException e) { 
        
            return false; 
        }
        catch(NullPointerException e) { 
        
            return false; 
        }
        
        return true;
    }
    
    public static boolean isDouble(String s) {
        
        try { 
            Double.parseDouble(s); 
        } 
        catch(NumberFormatException e) { 
        
            return false; 
        }
        catch(NullPointerException e) { 
        
            return false; 
        }
    
        return true;
    }
    
    /**
     *
     * @param filename
     */
    @Override
    public void readFile(String filename) throws myException{
        try{

            File file = new File(filename);

            if(file.length() < 3){

                throw new myException("Empty File");
            }
        
            input = new Scanner(new File (filename) );
            
            BufferedReader reader = null;
            
            reader = new BufferedReader(new FileReader(filename));
            
            String line = reader.readLine();
            
            for (String retval: line.split(" ")) {
                
                if(dataType == 1){
                    
                    if(isInteger(retval)){
 
						array.add((T)new Integer(retval));
                     }

                    else{

                        throw new myException("Bad data for int");

                    }

                }else if (dataType == 2){
                
					array.add((T)retval);
					                
                }else if (dataType == 3){
                    
                    if(isDouble(retval)){
                        
						array.add((T)new Double(retval));
                    }

                    else{
                        
                        throw new myException("Bad data for int");
                    }
                }
   
                countForNumGrams++;

            }

            peyir = new Pair[countForNumGrams-1];

            for(int z = 0 ; z < countForNumGrams-1 ; z++){

                peyir[z] = new Pair(array.get(z),array.get(z+1));

            }//pair oluşturuyor


            for (int i = 0 ; i < peyir.length; i++){

				if(mymap.containsKey(peyir[i]))
				
					mymap.put(peyir[i], mymap.get(peyir[i])+1);
					 
				else
					mymap.put(peyir[i], 1);

            }//end for

                    
            }//end try

        catch(FileNotFoundException fileNotFoundException){
            
            System.err.println("File could not found");        
        }
        catch(IOException ioException){
            
            System.err.println("File could not found");        
        }
        
        
        
    }//end of function
    
    @Override
    public int numGrams(){
        
        return countForNumGrams-1;
    }
    
    @Override
    public int numOfGrams(T var1,T var2){

		Integer result;
		
        result = mymap.get(new Pair(var1, var2));
        
        if(result == null)
        
			return 0;
			
		else
		
			return result.intValue();
    }
    
    @Override
    public String toString(){
        
        StringBuilder result = new StringBuilder();
        
        int maxValue=(Collections.max(mymap.values()));

        for(int max=maxValue ; max > 0 ; max--){

            for (Entry<Pair<T,T>,Integer> entry : mymap.entrySet()) {

                if (entry.getValue()==max) {

                    result.append("map.first : "+entry.getKey() + " num: " + entry.getValue() + "\n");    
                }
            }
        }
        return result.toString();
    }
    
}//end of class

File: 151044081_hw7/BigramMap.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <vector>
#include <exception>
#include "Bigram.h"
#include "BigramMap.h"
#include "Exception.h"
#include <sstream>


using namespace std;


template <class T>
void  BigramMap<T>::readFile(const string filename) throw (myException){

	ifstream file,file2;

	string line ;

	string temp;

	int a = 0 ;

	file.open(filename);

	file2.open(filename);

	string n;

	getline(file, n);

	istringstream k(n);

	k >> n;

	if(k.eof()){

		throw myException("EMPTY FILE!");
	}

	if(file.fail()){

		throw myException("File could not opened");


	}
	file.seekg(0);
	
	while(file>>temp){

		if(dataType == 1)

			if(badDataForint(temp)){

			}

			else{

				throw myException("Bad data for int file");
			}

		else if(dataType == 2){

			if(badDataFordouble(temp)){

			}
			else
				throw myException("Bad data for double file");
		}
	}

	getline(file2,line);

	if(line.length() == 0)

		throw myException("Empty file");

	T var ;

	istringstream iss(line);

	while(!iss.eof()){

		if(iss.fail())

			throw myException("Bad data");

		iss >> var;

		v.push_back(var);

		countForNumGrams++;

	}

	int s = 0 ;

	int b = 0 ;

	pair<T,T> peyir[countForNumGrams];

	for(b,s ; b < countForNumGrams-1, s < countForNumGrams-1; b++,s++){//pair olustur

		peyir[s].first = (v[b]);

		peyir[s].second = (v[b+1]);


	}
	for(int c = 0 ; c < countForNumGrams-1 ; c++ ){

		auto found = mymap.find(pair <T,T>(v[c],v[c+1]));



		if(found == mymap.end()){

			mymap.insert(make_pair(pair<T,T>(v[c],v[c+1]),1));//map'te yoksa map'e ekle
		}

		else{//map'te varsa value'yu 1 arttır

			auto f2 = found->first;

			found->second = ((found->second) +1);

			
		}
	}//end for


}//end function

template<class T>
int BigramMap<T>::numOfGrams(const T var1,const T var2){

	auto find = mymap.find(pair<T,T>(var1,var2));

	if(find != mymap.end()){

		auto f1 = find->first;

		auto f2 = find->second;

		return f2;

	}
	else{

		return 0;


	}


}

template<class T>
pair<T,T> BigramMap<T>::maxGrams(){

	auto it = mymap.begin();

	auto end = mymap.end();

	int max_value = it->second;

	auto str = it->first;

	for( ; it != end; ++it) {

    	if(it->second > max_value) {

        	max_value = it->second;

        	str = it->first;
    	}
    
	}

	return str;

}

template<class T>
int BigramMap<T>::numGrams() const {

	return countForNumGrams-1;
}

template<class T>
bool BigramMap<T>::badDataForint(string line){

	int i = 0;

	for(int j = 0 ; j < line.length(); j++){

		int a = line[i] - '0';

		if(a >= 0 && a <10){

			return true;
		}
		
		else{
			return false;
						
		}

		i++;
	}

}
template<class T>
bool BigramMap<T>::badDataFordouble(string line){
	int count = 0 ;
	int i = 0 ;

	for(int k = 0 ; k < line.length() ; k++){

		if(line[k] == '.'){

			count ++;
		}
	}

	for(int j = 0; j < line.length() ; j++){

		int a = line[i] - '0';

		if((a >= 0 && a <10) && count == 1)

			return true;

		else{

			return false ;
		}
						
		i++;
	}	

}
template<class T>
void BigramMap<T>::print(ostream& out){

	auto x = maxGrams();

	auto find = mymap.find(x);

	int value = find->second;

	for(value ; value >0 ; value--){

		for(auto y : mymap){

			if(y.second == value)

				out<<"first.first :"<<y.first.first<<" first.second :"<<y.first.second<<" second :"<<y.second<<endl;
		}
	}

}